### DISCUSSION TITLE
Double Checking How Kubernetes Works

### DISCUSSION THREAD

**Discussion by catvec** [qdm12/gluetun#1762] [Q&A] [answered]
In the wiki there are instructions for using `network_stack` to link a container to the VPN container. I was wondering if there is anything similar that needs to happen for Kubernetes when running Glueten as a sidecar container in a pod.

I want to be sure before I start running traffic on my Kubernetes server, so I don't accidentally leak any data out through a non VPN route.

My understanding is that containers in the same pod share the same networking namespace. Which is essentially what the `network_mode: "service:gluten"` option does? So Gluten should just work as a Kubernetes pod sidecar container?

Additional thought: Has anyone tried using Kubernetes Network Policies on a pod to ensure that traffic only goes through the VPN?

**Comment by renannprado** [comment]
I'm also interested. Have you find the answer @Noah-Huppert? I hanve't done that yet, but my plan is to apply policies with Calico.

**Comment by catvec** [comment, marked as answer]
@renannprado Oh shoot sorry for not updating this thread with my solution. Basically what I learned is that within a Kubernetes pod all stuff related networking is shared between containers (at least using Digital Ocean managed k8s which I believe uses Cilium). This means if you run Glutun as one of the containers in your pod, other containers will only be able to send traffic via the VPN.

I used Kubernetes network policies just as an extra safety net in case the VPN shut down. 

[Here is a link to the Kustomize stack I made which runs qBittorrent and Glutun](https://github.com/Noah-Huppert/funkyboy.zone/tree/5cc4ecee2b3d8d02c0b7c2c174f7cd713b4ed59e/kubernetes/base/media-server/bases/qbittorrent). [This is specifically how I ran Glutun](https://github.com/Noah-Huppert/funkyboy.zone/blob/5cc4ecee2b3d8d02c0b7c2c174f7cd713b4ed59e/kubernetes/base/media-server/bases/qbittorrent/resources/deployment.yaml#L49) and [here is where I setup a network policy only allowing VPN traffic](https://github.com/Noah-Huppert/funkyboy.zone/blob/5cc4ecee2b3d8d02c0b7c2c174f7cd713b4ed59e/kubernetes/base/media-server/bases/qbittorrent/resources/network-policy.yaml). I obtained the IPs in the network policy using [these steps](https://github.com/Noah-Huppert/funkyboy.zone/blob/5cc4ecee2b3d8d02c0b7c2c174f7cd713b4ed59e/kubernetes/base/media-server/README.md#qbittorrent-pod-network-policies).

**Reply by renannprado** [reply]
Wow. That's some good stuff. I'll have a look, thanks!

**Reply by enzanto** [reply]
Hi! i just stumbled over this docker image, and had to dive straight in to learning how to apply it in my cluster properly. From kubernetes v1.28 it should be used as a native sidecar. As long as the sidecar is active it blocks the traffic from the main container (even before it connects to vpn). I did a short write up [here](https://nerdiverset.no/k8s-native-sidecar-with-vpn/). I am planning to also make a pull request in to the wiki here with the same info.

**Reply by mstaffa** [reply]
In your write-up you mention that you use `DNS_KEEP_NAMESERVER` to keep access to the Kubernetes DNS resolver. While that works, you're likely to leak DNS requests outside the VPN as mentioned in the gluetun documentation [here](https://github.com/qdm12/gluetun-wiki/blob/main/setup/options/dns.md).

If that suits your needs then great! To anyone else reading this comment though, be warned that leaked DNS traffic can loudly announce whatever you want your VPN to keep private. There is an [open issue](https://github.com/qdm12/gluetun/issues/1839) I've commented on that addresses this.
